prow/crier/controller_test.go

```diff
@@ -0,0 +1,183 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package crier
+
+import (
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/test-infra/prow/apis/prowjobs/v1"
+	"k8s.io/test-infra/prow/client/clientset/versioned/fake"
+	v12 "k8s.io/test-infra/prow/client/listers/prowjobs/v1"
+	"k8s.io/test-infra/prow/kube"
+	"sync"
+	"testing"
+	"time"
+)
+
+const (
+	testTimeLimit = 1 * time.Second
+)
+
+//Fake Reporter
+//Sets: Logic; Some series of test jobs should be reported, while others shouldn't be
+//Asserts: The correct jobs have Report called on them
+type failureReporter struct {
+	passing *bool
+	failed  *bool
+}
+
+func (f failureReporter) Report(pj *v1.ProwJob) ([]*v1.ProwJob, error) {
+	if pj.Status.State == v1.FailureState {
+		*f.passing = true
+	} else {
+		*f.failed = true
+	}
+	return nil, nil
+}
+
+func (f failureReporter) GetName() string {
+	return "failureReporter"
+}
+
+func (f failureReporter) ShouldReport(pj *v1.ProwJob) bool {
+	return pj.Status.State == v1.FailureState
+}
+
+//Fake Informer
+//Sets: The Prow Job Test Cases
+//Asserts: The Controller registers for the queue to be populated
+type fakeInformer struct {
+	isRun        *bool
+	isRegistered *bool
+}
+
+func (f fakeInformer) Get(name string) (*v1.ProwJob, error) {
+	var pj *v1.ProwJob = nil
+	switch name {
+	case "TestJob1":
+		pj = &v1.ProwJob{
+			Spec: v1.ProwJobSpec{
+				Job: "TestJob1",
+			},
+			Status: v1.ProwJobStatus{
+				State: v1.TriggeredState,
+			},
+		}
+	case "TestJob2":
+		pj = &v1.ProwJob{
+			Spec: v1.ProwJobSpec{
+				Job: "TestJob2",
+			},
+			Status: v1.ProwJobStatus{
+				State: v1.FailureState,
+			},
+		}
+	}
+
+	if pj == nil {
+		panic("TODO: return an error")
```

Comment:
fix the TODO? :-)

Reply:
Oh, whoops, I was certain I had fixed that

I think it might be best to have this function "return pj, nil" even if pj isn't assigned.
"Get" returning a nil pointer when it can't find its target doesn't seem that unusual.

Another option is to fail the test, but I don't know if that's the right thing to do in this case.